contracts/Market.sol

```diff
@@ -941,144 +772,72 @@ contract Market is
-   * specification of the fee percentage that was paid off-chain.. If the specified supplier does not have enough carbon
-   * removals for sale to fulfill the order the transaction will revert.
-   * @dev See [here](https://docs.openzeppelin.com/contracts/4.x/api/token/erc20#IERC20-approve-address-uint256-) for
-   * more. The purchaser must have granted approval to this contract to authorize this market to transfer their
-   * supported ERC20 tokens to complete the purchase. A certificate is issued by the Certificate contract
-   * to the specified recipient and the ERC20 tokens are distributed to the supplier(s) of the carbon removal as well as
-   * potentially to the RestrictedNORI contract that controls any restricted portion of the ERC20 owed to the supplier.
-   *
-   * ##### Requirements:
-   *
-   * - Can only be used when this contract is not paused.
-   * - Can only be used when the caller has the `MARKET_ADMIN_ROLE` role.
-   * - Can only be used when the specified supplier has enough carbon removals for sale to fulfill the order.
-   * - Can only be used if this contract has been granted approval to spend the purchaser's ERC20 tokens.
-   * @param recipient The address to which the certificate will be issued.
-   * @param purchaser The address that will pay for the removals and has granted approval to this contract
-   * to transfer their ERC20 tokens.
-   * @param amount The total purchase amount in ERC20 tokens. This is the total number of removals being
-   * purchased, scaled by the price multiple.
-   * @param supplier The only supplier address from which to purchase carbon removals in this transaction.
-   * @param customFee The custom fee percentage that was paid to Nori, as an integer, specified here for
-   * inclusion in emitted events.
-   * @param customPriceMultiple The custom price to be used for this transaction.
-   */
-  function swapFromSupplierWithoutFeeSpecialOrder(
-    address recipient,
-    address purchaser,
-    uint256 amount,
-    address supplier,
-    uint256 customFee,
-    uint256 customPriceMultiple
-  ) external whenNotPaused onlyRole(MARKET_ADMIN_ROLE) {
-    uint256 currentPrice = _priceMultiple;
-    _priceMultiple = customPriceMultiple;
-    _validateCertificateAmount({amount: amount});
-    (
-      uint256 countOfRemovalsAllocated,
-      uint256[] memory ids,
-      uint256[] memory amounts,
-      address[] memory suppliers
-    ) = _allocateRemovalsFromSupplier({
+      // case vintage-specific fulfillment and supplier-specific fulfillment
+    } else if (vintages.length > 0 && supplier != address(0)) {
+      (
+        countOfRemovalsAllocated,
+        ids,
+        amounts,
+        suppliers
+      ) = _allocateRemovalsFromSupplierSpecificVintages({
         certificateAmount: amount,
+        vintages: vintages,
         supplier: supplier
       });
+    }
+    // case no restrictions on fulfillment
+    else {
```

Comment:
yes actually -- and the only place I needed to add this was in the implementation of `allocateRemovalsSpecificVintages` ... in the happy path case this is taken care of by `validateSupply`, and in the supplier specific cases that logic already exists in the allocation logic because there is some pre-loop overhead that looks at the supplier's available removals ahead of time.  In this way, each of the 4 possible fulfillment paths uses the information that is efficiently available to it to abort as early as possible without doing a bunch of extra work.